include/kernel/faulthandler.h

```diff
@@ -7,11 +7,29 @@
 #pragma once
 
 #include <object.h>
+#ifdef CONFIG_KERNEL_MCS
+#include <kernel/sporadic.h>
+#endif
 
 #ifdef CONFIG_KERNEL_MCS
 static inline bool_t validTimeoutHandler(tcb_t *tptr)
 {
-    return cap_get_capType(TCB_PTR_CTE_PTR(tptr, tcbTimeoutHandler)->cap) == cap_endpoint_cap;
+    cap_t handlerCap = TCB_PTR_CTE_PTR(tptr, tcbTimeoutHandler)->cap;
```

Comment:
Can you do any optimisations here with the knowledge that the timeout handler cap is installed by the kernel and cannot be deleted? I'm thinking you can set a pointer in the TCB to the endpoint object then directly check it's state without a lookup if that pointer is set.

Reply:
I don't think so. If the cap gets destroyed via revocation then there is no way to determine that the cap is being used by the TCB as a timeout EP, so the TCB would end up with a stale pointer.

Reply:
Aha, but recovation only occurs when all caps to an object are deleted. Users cannot delete the cap in the TCB Cnode without deleting the TCB. So if the TCB exists, any cap in the Cnode is a cap to the object, which continues to exist.

Reply:
Except in the case of revocation, which allows copies of caps to be removed without needing direct access to the CNode in which they reside. Revoking the untyped for the EP or the original EP capability can removed it from the TCB, so we can't statically guarantee the pointer implies presence of the capability easily.

Reply:
So when you call finaliseCap you should be able to undo the link?

Reply:
Yes, I'm well aware what revoke does.

Reply:
When you finaliseCap you don't have a reference to the TCB.

Reply:
If you can get to the TCB CNode you can get the TCB pointer easily. I don't recall if you can do that from revoke but there's likely a way to do it. You could also consider a link between the TCP and EP objects, but I suspect that's not required. 

Anyway, you don't seem interested in pursuing this idea, and I'm done offering suggestions.

Reply:
I think you *can* do it, knowing that if you know the cap is specifically being used as the timeout endpoint of a TCB, you can determine where the CNode starts and where the TCB is based on knowing which slot it is in, but in the general case you can't determine the location of a CNode from a capability slot.

Reply:
I also think that while the overheads of the checks involved here should be optimised, that they can be addressed separately as it seems that moving the necessary checks elsewhere and guaranteeing the necessary invariants required to make them correct such that the check here is lightweight are going to add significant complexity in their own right.

Reply:
Once something is verified it's very hard to change. Also what are the impact of this change on the scheduling benchmarks that use timeout faults?